### PR TITLE
Stop eating panics

### LINKED ISSUE #28365: Remove the defer calls that catch panics
They tend to hide bugs in production: software crashes make it really clear when there is a problem; catching panics papers over the issue and makes it more difficult for monitoring systems to catch failures. 

/cc @lavalamp @krousey

### PR DESCRIPTION
**Pull request by lavalamp**
```
After this change, by default Kubernetes components will stop handling panics and actually crash. All Kubernetes components should be run by something that actively restarts them. This is true of the default setups, but those with custom environments may need to double-check.

If necessary, one can change this behavior back to the previous behavior by setting pkg/util/runtime's `ReallyPanic` to false.
```

Fixes #28365

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28800)

### CODE REVIEW

**Comment by lavalamp** [comment]
looks like some tests need attention.

**Comment by roberthbailey** [comment]
lgtm other than fixing the unit/integration tests.

**Comment by lavalamp** [comment]
Hm, I'm not sure if we want to turn these off for kubelet. @yujuhong, would anything bad happen if kubelet started actually crashing when it panics?

**Comment by yujuhong** [comment]
> Hm, I'm not sure if we want to turn these off for kubelet. @yujuhong, would anything bad happen if kubelet started actually crashing when it panics?

Uh...you can say that in most of the cluster setup, there will be some babysitter daemon restarting kubelet. I am not sure if it's true for all setup though. On the other hand, I've seen kubelet panic'd and then the panic got handled by the util, leading to starting multiple instances of the same components. It may be safer/easier to debug to let kubelet crash.

**Comment by lavalamp** [comment]
OK, this should make the tests pass.

**Comment by lavalamp** [comment]
Thanks, @yujuhong -- I think I'll just disable it everywhere like I originally planned then.

fyi @smarterclayton in case this affects OpenShift

**Comment by yujuhong** [comment]
/cc @kubernetes/sig-node 

After this PR is merged, if kubelet panics, it will crash.

File: pkg/util/runtime/runtime.go [hunk 1]
```diff
@@ -42,6 +50,8 @@ func HandleCrash(additionalHandlers ...func(interface{})) {
 		for _, fn := range additionalHandlers {
 			fn(r)
 		}
+		// Actually proceed to panic.
+		panic(r)
```
Comment (smarterclayton) [inline pkg/util/runtime/runtime.go]: That's terrifying.

File: pkg/util/runtime/runtime.go [hunk 2]
```diff
@@ -23,14 +23,22 @@ import (
...
+// defer.  Additional context-specific handlers can be provided, and will be
+// called in case of panic.  HandleCrash actually crashes, after calling the
+// handlers and logging the panic message.
+//
+// TODO: remove this function. We are switching to a world where it's safe for
+// apiserver to panic, since it will be restarted by kubelet. At the beginning
+// of the Kubernetes project, nothing was going to restart apiserver and so
+// catching panics was important. But it's actually much simpler for montoring
+// software if we just exit when an unexpected panic happens.
 func HandleCrash(additionalHandlers ...func(interface{})) {
```
Comment (smarterclayton) [inline pkg/util/runtime/runtime.go]: Well, we want the handlers to be called.  The handlers allow us to report to external systems (on my long delayed list is to upstream the sentry code that reports all panics to a remote system).

**Comment by smarterclayton** [comment]
Somewhat of a big change.  Doesn't this mean any panic someone can reproduce can be a denial of service attack against any component of the infra?  I do like making these more visible, but for API servers exposed to multi-tenant users this is riskier.

Can we flip the default so that someone can disable this if they disagree?

File: pkg/util/runtime/runtime.go (hunk 2 above)
Comment (lavalamp) [inline pkg/util/runtime/runtime.go]: The handlers are still called.

**Comment by lavalamp** [comment]
Updated. ReallyCrash boolean works again.

**Comment by smarterclayton** [comment]
Thanks, looks good to me aside from whatever the failures are in mesos tests.

**Comment by lavalamp** [comment]
I'm pretty sure the remaining mesos test error is actually a legit test error that was being papered over by us eating this panic. @kubernetes/kube-mesos @karlkfi Shall I disable the test until you have time to look at it? Or can you suggest a fix?

**Comment by jdef** [comment]
@lavalamp please disable the test for now - thanks

**Comment by lavalamp** [comment]
Thanks @jdef. Test disabled.

**Comment by lavalamp** [comment]
Argh, go vet doesn't like "dead code", one sec...

File: contrib/mesos/pkg/executor/executor_test.go [hunk 3]
```diff
@@ -380,8 +380,10 @@ func TestExecutorInitializeStaticPodsSource(t *testing.T) {
 // and Kamikaze.  When a task is lost, the executor needs to clean up
 // its state.  When a Kamikaze message is received, the executor should
 // attempt suicide.
-func TestExecutorFrameworkMessage(t *testing.T) {
-	// create and start executor
+//
+// TODO(jdef): This test started failing when panic catching was
+// disabled.
+func DisabledTestExecutorFrameworkMessage(t *testing.T) {
```
Comment (euank) [inline contrib/mesos/pkg/executor/executor_test.go]: I think leaving the test name alone and putting `t.Skip` at the top is a more idiomatic way to skip it and dodge the dead code go vet warnings

Comment (lavalamp) [inline contrib/mesos/pkg/executor/executor_test.go]: ah cool, done

**Comment by roberthbailey** [comment]
I'm going to mark this as `release-note-action-required` since it's possible to install k8s in such a way that the apiserver isn't babysat by the kubelet.

**Comment by lavalamp** [comment]
@roberthbailey Yes, good call. And do we want to cherry-pick?

**Comment by lavalamp** [comment]
@k8s-bot test this issue #IGNORE (github flake?)

**Comment by roberthbailey** [comment]
I'd like to cherry pick into 1.3, barring objections from the community.

**Comment by lavalamp** [comment]
all the tests pass, I'm going to interpret Robert & Clayton's comments as a collective LGTM.

**Comment by smarterclayton** [comment]
Yes, it was LGTM

**Comment by thockin** [comment]
Did you want to offer a cherry-pick for next cut?

**Comment by fabioy** [comment]
@lavalamp Ping. Could you create a cherrypick PR?

**Comment by lavalamp** [comment]
I'm... concerned about pushing this into prod, since apparently controller manager deletes and recreates a bunch of things it ought not when it restarts.

**Comment by roberthbailey** [comment]
I'm ok waiting for this to land in the 1.3 branch after 1.3.3 to give it a longer bake time (both at head and in the release branch).

**Comment by fabioy** [comment]
Since 1.4 is close to release, I guess we won't be taking this on the 1.3 branch, so removing the cherrypick labels.

**Comment by spiffxp** [comment]
@lavalamp @roberthbailey this PR has a release-note-action-required label on it, but I can't tell that the action is something an end-user of kubernetes can do, outside of kubelet's `--really-crash-for-testing` flag

can you clarify whether this is worth calling out as an action required in the 1.4.0 release notes?

**Comment by lavalamp** [comment]
@spiffxp the change message says:

> After this change, by default Kubernetes components will stop handling panics and actually crash. All Kubernetes components should be run by something that actively restarts them. This is true of the default setups, but those with custom environments may need to double-check.

This is the action that is needed.

**Comment by lavalamp** [comment]
The action is targeted at cluster admins. Most action-required items are, I think. End users of kubernetes should not need to take many actions.

**Comment by spiffxp** [comment]
@lavalamp agree on clusters ops audience, hence my confusion, skipped over that first line, thanks for pointing me in the right direction

**Pull request merged**